### DISCUSSION TITLE
How to calculate the handling fee?

### DISCUSSION THREAD

**Discussion by Smilestate** [polkascan/py-substrate-interface#300] [Q&A] [answered]
I ran into a problem. When I transferred money, the fee received through get_payment_info was incorrect. How to accurately obtain its handling fee.
The second question is that there is a tip when transferring money, so how do you calculate the tip and how to enter it in the transfer method?

**Comment by arjanz** [comment, marked as answer]
> I ran into a problem. When I transferred money, the fee received through get_payment_info was incorrect. How to accurately obtain its handling fee. 

I ran a test transfer with the following call:

```python
keypair = Keypair.create_from_uri('//Alice', ss58_format=42)

balance_info = substrate.query(
    module='System',
    storage_function='Account',
    params=[keypair.ss58_address],
)

print('Balance info', balance_info.value)

call = substrate.compose_call(
    call_module='Balances',
    call_function='transfer_keep_alive',
    call_params={
        'dest': keypair.ss58_address,
        'value': 600000000000
    }
)

info = substrate.get_payment_info(call, keypair)
# result: {'class': 'normal', 'partialFee': 169117900, 'weight': 176532000}
```

So the fee estimate of this call is: 169117900 plancks.

After submitting this extrinsic the fee (which is defined by the sum of the events `Treasury.Deposit` and `Balances.Deposit`) is: 170117159 plancks. So the estimate is actually quitte accurate (~0.6% less), I think it is impossible to get a predication exactly matching with the real transaction because the state changes constantly. 

![Screenshot 2021-10-26 at 11 59 50](https://user-images.githubusercontent.com/5286904/138857084-0a996aed-6c70-4426-a694-8e2fd212bc1d.png)


In conclusion, I don't see anything wrong with this function, but of course I encourage you to proof me wrong ;)

> The second question is that there is a tip when transferring money, so how do you calculate the tip and how to enter it in the transfer method?

The (optional) tip that you can specify when creating an extrinsic is really treated like a tip, the validator will first select the extrinsic (=transaction) that has the most tip, then transactions that has no tip. So basically this is only useful when the transaction pool is full and you want to pay extra to be processed first. So at the moment for Polkadot/Kusama this is not necessary yet, hundreds of transactions per block is no problem for Substrate.

**Comment by arjanz** [comment]
You can specify the tip when creating the extrinsic: https://polkascan.github.io/py-substrate-interface/#substrateinterface.SubstrateInterface.create_signed_extrinsic

**Comment by Smilestate** [comment]
> 您可以在创建外部时指定提示：[https](https://polkascan.github.io/py-substrate-interface/#substrateinterface.SubstrateInterface.create_signed_extrinsic) : [//polkascan.github.io/py-substrate-interface/#substrateinterface.SubstrateInterface.create_signed_extrinsic](https://polkascan.github.io/py-substrate-interface/#substrateinterface.SubstrateInterface.create_signed_extrinsic)

thank you!
